microsetta_admin/metadata_util.py

```diff
@@ -1,141 +1,258 @@
-        for group in template["survey_template_text"]["groups"]:
-            for field in group["fields"]:
-                visit_field(final_map, field)
-
-    return final_map
+
+        fetched.append(barcode_metadata)
+
+    if len(fetched) == 0:
+        error_report.append({"error": "No metadata was obtained"})
+        df = pd.DataFrame()
+    else:
+        df = _to_pandas_dataframe(fetched)
+
+    return df, error_report
+
+
+def _to_pandas_dataframe(metadatas):
+    """Convert the raw barcode metadata into a DataFrame
+
+    Parameters
+    ----------
+    metadatas : list of dict
+        The raw metadata obtained from the private API
+
+    Returns
+    -------
+    pd.DataFrame
+        The fully constructed sample metadata
+    """
+    transformed = []
+    multiselects = set()
+
+    for metadata in metadatas:
+        as_series, observed_multiselect = _to_pandas_series(metadata)
+        transformed.append(as_series)
+        multiselects.update(observed_multiselect)
+
+    df = pd.DataFrame(transformed)
+    df.index.name = 'sample_name'
+
+    # update reponse where someone did not check a multi-select box
+    for multiselect in multiselects:
+        nulls = df[multiselect].isnull()
+        df.loc[nulls, multiselect] = 'false'
+
+    # fill in any other nulls that may be present in the frame
+    # as could happen if not all individuals took all surveys
+    df.fillna('Missing: not provided', inplace=True)
+
+    return df
+
+
+def _to_pandas_series(metadata):
+    """Convert the sample metadata object from the private API to a pd.Series
+
+    Parameters
+    ----------
+    metadata : dict
+        The response object from a query to fetch all sample metadata for a
+        barcode.
+
+    Returns
+    -------
+    pd.Series
+        The transformed responses
+    set
+        Observed multi-selection responses
+    """
+    name = metadata['sample_barcode']
+    hsi = metadata['host_subject_id']
+    source_type = metadata["source_type"]
+
+    sample_detail = metadata['sample']
+    collection_timestamp = sample_detail['datetime_collected']
+
+    if source_type == 'human':
+        sample_type = sample_detail['site']
+        sample_invariants = HUMAN_SITE_INVARIANTS[sample_type]
+    elif source_type == 'animal':
+        sample_type = sample_detail['site']
+        sample_invariants = {}
+    else:
+        sample_type = sample_detail['source']['description']
+        sample_invariants = {}
+
+    values = [hsi, collection_timestamp]
+    index = ['HOST_SUBJECT_ID', 'COLLECTION_TIMESTAMP']
+
+    # TODO: denote sample projects
+    observed_multiselect = set()
+    for survey in metadata['survey_answers']:
+        for shortname, answer in survey['response'].values():
+            if isinstance(answer, list):
+                for selection in answer:
+                    specific_shortname = _build_col_name(shortname, selection)
+                    values.append('true')
```

Comment:
Why the string 'true' rather than the boolean value?  Where do we instead add 'false' when an option was not selected?

Reply:
`false` and `true` are the predominant values currently used in Qiita, and a perpetual need is to redo all that metadata. To make it a little nastier, this isn't a true boolean column as a None value is also allowed.

Reply:
Agh, but how can it write None to this table?  Either it wrote 'true', or it is overridden here to write 'false', in order to properly write None, this code needs to keep track of whether or not that barcode has taken that survey.

Reply:
The "false" values are filled in after the construction of the dataframe. Sorry for the confusion, a logical null value is allowed in the bigger picture of AG metadata which includes things like blank samples.

But, the notion of "false" here gets really difficult in part because of the voluntary nature of the survey. A lack of a response does not strictly mean false.